R/step_adjust_latency.R

```diff
@@ -0,0 +1,273 @@
+        shift_cols = NULL,
+        prefix = prefix,
+        default = default,
+        keys = epi_keys(recipe),
+        skip = skip,
+        id = id
+      )
+    )
+  }
+
+step_adjust_latency_new <-
+  function(terms, role, trained, as_of, latency, shift_cols, prefix, time_type, default,
+           keys, method, skip, id) {
+    step(
+      subclass = "adjust_latency",
+      terms = terms,
+      role = role,
+      method = method,
+      trained = trained,
+      as_of = as_of,
+      latency = latency,
+      shift_cols = shift_cols,
+      prefix = prefix,
+      default = default,
+      keys = keys,
+      skip = skip,
+      id = id
+    )
+  }
+
+# lags introduces max(lags) NA's after the max_time_value.
+# TODO all of the shifting happens before NA removal, which saves all the data I might possibly want; I should probably add a bit that makes sure this operation is happening before NA removal so data doesn't get dropped
+#' @export
+prep.step_adjust_latency <- function(x, training, info = NULL, ...) {
+  if ((x$method == "extend_ahead") && (!("outcome" %in% info$role))) {
+    cli::cli_abort(glue::glue(c('If `method` is `"extend_ahead"`, then a step ",
+          "must have already added an outcome .')))
+  } else if (!("predictor" %in% info$role)) {
+    cli::cli_abort('If `method` is `"extend_lags"` or `"locf"`, then a step ",
+"must have already added a predictor.')
+  }
+
+  sign_shift <- get_sign(x)
+  # get the columns used, even if it's all of them
+  terms_used <- x$columns
+  if (length(terms_used) == 0) {
+    terms_used <- info %>%
+      filter(role == "raw") %>%
+      pull(variable)
+  }
+  # get and check the max_time and as_of are the right kinds of dates
+  if (is.null(x$as_of)) {
+    as_of <- set_asof(training, info)
+  } else {
+    as_of <- x$as_of
+  }
+
+  # infer the correct columns to be working with from the previous
+  # transformations
+  shift_cols <- get_shifted_column_tibble(
+    x$prefix, training, terms_used,
+    as_of, x$latency, sign_shift, info
+  )
+
+  if ((x$method == "extend_ahead") || (x$method == "extend_lags")) {
+    # check that the shift amount isn't too extreme
+    latency <- max(shift_cols$latency)
+    time_type <- attributes(training)$metadata$time_type
+    if (
+      (grepl("day", time_type) && (latency >= 10)) ||
+        (grepl("week", time_type) && (latency >= 4)) ||
+        ((time_type == "yearmonth") && (latency >= 2)) ||
+        ((time_type == "yearquarter") && (latency >= 1)) ||
+        ((time_type == "year") && (latency >= 1))
+    ) {
+      cli::cli_warn(c(
+        "!" = glue::glue(
+          "The shift has been adjusted by {latency}, ",
+          "which is questionable for it's `time_type` of ",
+          "{time_type}"
+        ),
+        "i" = "input ahead: {shift_cols$shifts[[i_latency]]}",
+        "i" = "shifted ahead: {shift_cols$effective_shift[[i_latency]]}",
+        "i" = "max_time = {max_time} -> as_of = {as_of}"
```

Comment:
I've been burned by it multiple times when using my external version of latency adjustment, and the error that came up as a result wasn't very comprehensible. I'm not sure I'd ever use `step_adjust_latency` without also using `check_latency`.

We could also set the levels to something more extreme, e.g. latency > 365 for day, or 48 for week etc, since it was usually having `as_epi_df` defaulting to the `as_of` being today that was causing me problems with retrospective forecasting